### DISCUSSION TITLE
proposal: `@oxc/jest` package

### DISCUSSION THREAD

**Discussion by DonIsaac** [oxc-project/oxc#4247] [Feature Request]
The [Jest](https://jestjs.io/) testing framework supports transforming code via transform plugins, such as [`babel-jest`](https://www.npmjs.com/package/babel-jest), [`ts-jest`](https://www.npmjs.com/package/ts-jest), and [`@swc/jest`](https://swc.rs/docs/usage/jest). It would be a large quality-of-life improvement to use oxc to transform files instead of any of the above packages. The transformer is progressing quite well, and making it usable in Jest could help with the adoption of oxc in general.

Unfortunately, Jest only ships with [`experimental support for ESM`](https://jestjs.io/docs/ecmascript-modules), so transpilation to CJS is a must-have for this to work properly. I took a crack at this about two weeks ago. Within about 2 hours I got about 80% of the way there, with this missing plugin being the only remaining major blocker.

## Goals
- Typescript/ESM transpilation into CJS
- Should support transpilation to the most recent Node LTS version
- Should work out-of-the-box, using fields from `package.json`, `tsconfig.json`, [`.babelrc`](https://babeljs.io/docs/config-files), etc to configure transpilation correctly

## Tasks
- [ ] feat: Expose JS bindings for `oxc_transform` as an NPM package, similar to `@oxc/parser`
- [ ] feat(transform): support `@babel/plugin-transform-modules-commonjs` plugin
- [ ] feat: add `@oxc/jest` NPM package
- [ ] docs: add setup and usage instructions for `@oxc/jest` to the website.

## Considerations
- The SWC project has `@swc/jest` in a [separate git repo](https://github.com/swc-project/pkgs/tree/main/packages/jest) from their [core package](https://github.com/swc-project/swc/tree/main/packages/core). Should we do the same thing?
- Below are what I see as the major pitfalls of `@swc/jest` that prevent widespread adoption:
  1. `@swc/jest` unconditionally hoists package imports, which breaks module mocking. This behavior is undocumented, and attempts to fix it by setting the appropriate config property fail as the package always overrides this user-provided value.
  2.  `.swcrc` does not appear to be respected; I've only ever gotten configuration to work by passing a config option explicitly.
  3. `@swc/jest` does not export type definitions for its config, and the various settings are completely undocumented
  4. External source maps do not work correctly, only inlined ones do. This messes up debugging and code coverage.

**Comment by Boshen** [comment]
My only concern is the maintenance cost, the core team will not have the time to maintain this :-(

I'm fine with shipping all the puzzles pieces have a separate repo that glues things together with its own release pipeline.

**Comment by DonIsaac** [comment]
> My only concern is the maintenance cost, the core team will not have the time to maintain this :-(
> 
> I'm fine with shipping all the puzzles pieces have a separate repo that glues things together with its own release pipeline.

Once transformation is exposed as an npm package, creating a jest plugin package is relatively trivial. `@swc/jest` is [one file and about 200 lines of code](https://github.com/swc-project/pkgs/blob/main/packages/jest/index.ts). I'd be happy to maintain it myself. Would you be willing to create a git repo for it within `oxc-project` and give me the permissions needed to maintain it, or should I create it myself?

**Comment by Boshen** [comment]
> should I create it myself

Yes please, ping me if you don't have all the permissions.

**Comment by DonIsaac** [comment]
https://github.com/oxc-project/oxc-jest
